### ISSUE TITLE
Incorrect plan generated for inequality joins and roll up

### ISSUE THREAD

**Issue by kuruv** [greenplum-db/gpdb#1342]
Using GPDB 4.3.10 with Optimizer PQO version 1.677

The following query 

select 
   a.member
   ,min(a.transdate)
from core.trans as a
join custlist as b 
   on a.member=b.member 
   and a.transdate between '2015-01-01' and current_timestamp
where a.transdate>b.min_trans_date
group by a.member;

generates query plan whose deepest three explain line items are - 

->  Partition Selector for trans (dynamic scan id: 1)  (cost=10.00..100.00 rows=4 width=4) 
	 Filter: trans.bank > custlist.min_trans_date 
	 ->  Broadcast Motion 32:32  (slice1; segments: 32)  (cost=0.00..499.69 rows=1257274 width=18) 
		   ->  Table Scan on custlist  (cost=0.00..432.71 rows=39290 width=18) 

Obviously, trans.bank > custlist.min_trans_date is a wrong operation. Also, when the code is run, it throws a high VMEM usage error.

The error goes away if -
1. The group by is removed
2. A clause "or a.transdate is null" is added to where (In this case, it is essentially the same query and the results are given in 20 sec)

**Comment by vraghavan78** [comment]
Hi @kuruv, this is a known explain bug. The plan is correct but the expression under the partition selector is printed incorrectly. Can you send me the DDL and we can fix this.

**Comment by kuruv** [comment]
Hi @vraghavan78, please find the DDLs attached. Forgive me if I am wrong, but if the plan is correct, its execution should not be throwing the high VMEM usage error, right?

Following is the error I get -

NOTICE:  Canceling query because of high VMEM usage. Used: 15180MB, available 1704MB, red zone: 15358MB (runaway_cleaner.c:135)  (seg22 slice2 app2060:40006 pid=34820) (cdbdisp.c:1320)

Following is the memory statistics with the "or a.transdate is null" clause - 

Slice statistics:
  (slice0)    Executor memory: 569K bytes.
  (slice1)    Executor memory: 1849K bytes avg x 32 workers, 1900K bytes max (seg17).
  (slice2)    Executor memory: 55170K bytes avg x 32 workers, 55177K bytes max (seg1).  Work_mem: 2186K bytes max.
  (slice3)    Executor memory: 34356K bytes avg x 32 workers, 34356K bytes max (seg0).
Statement statistics:
  Memory used: 1048576K bytes

[custlist.DDL.txt](https://github.com/greenplum-db/gpdb/files/599181/custlist.DDL.txt)
[core.trans.DDL.txt](https://github.com/greenplum-db/gpdb/files/599182/core.trans.DDL.txt)

**Comment by vraghavan78** [comment]
Thanks from the DDL. 

Sorry, I did not completely answer your original post. There are two issues here as I see (I will confirm when I get in tomorrow morning):
* Wrong predicate which could be the explain issue that we know (will confirm tomorrow)
* Execution issue. <--- This could be because of a bad plan that PQO generated. 

I saw that DDL, thank you for that, but I forgot to ask you the stats so we can generate the same plan as you have. This can be done in one of the following ways:
* dump the GPSD of the database
* minirepro https://discuss.pivotal.io/hc/en-us/articles/217976508-How-to-collect-DDL-and-statistics-information-using-the-minirepro-utility

**Comment by vraghavan78** [comment]
@kuruv I can confirm that the explain issue is the known issue I talked earlier about. Here is the details:

Consider a smaller query

```sql
explain 
select * from core.trans as a join custlist as b
on a.member=b.member and a.transdate between '2015-01-01' and current_timestamp
where a.transdate>b.min_trans_date;
```

The explain of the same is as follows:

```
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..862.00 rows=1 width=380)
   ->  Hash Join  (cost=0.00..862.00 rows=1 width=380)
         Hash Cond: trans.member = custlist.member
         Join Filter: trans.transdate > custlist.min_trans_date
         ->  Dynamic Table Scan on trans (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=328)
               Filter: transdate >= '2015-01-01 00:00:00'::timestamp without time zone AND transdate <= '2016-11-18 10:51:51.614599-08'::timestamp with time zone
         ->  Hash  (cost=100.00..100.00 rows=34 width=4)
               ->  Partition Selector for trans (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
                     Filter: trans.transdate > custlist.min_trans_date
                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=52)
                           ->  Table Scan on custlist  (cost=0.00..431.00 rows=1 width=52)
 Settings:  optimizer=on
 Optimizer status: PQO version 1.687
(13 rows)
```

The partition selector filter is the same as what you saw.

```
Filter: trans.transdate > custlist.min_trans_date
```

When you look at the minidump for this query you will see the DXL element for the printable fllter is correct.

```
            <dxl:PrintableFilter>
              <dxl:Comparison ComparisonOperator="&gt;" OperatorMdid="0.2064.1.0">
                <dxl:Ident ColId="13" ColName="transdate" TypeMdid="0.1114.1.0"/>
                <dxl:Ident ColId="49" ColName="min_trans_date" TypeMdid="0.1114.1.0"/>
              </dxl:Comparison>
            </dxl:PrintableFilter>
```

The explain bug is in the CTranslatorDXLToPlStmt. This only affects the display and not the condition.

We will fix this issue soon.

**Comment by vraghavan78** [comment]
@kuruv since you are using the enterprise edition of GPDB could you channel the discussion via Pivotal GSS so we can appropriately prioritize it.

**Comment by kuruv** [comment]
@vraghavan78 , please find attached attached the minirepro's output for both the codes:
[code2_querystats.txt](https://github.com/greenplum-db/gpdb/files/603107/code2_querystats.txt)
[code1_querystats.txt](https://github.com/greenplum-db/gpdb/files/603108/code1_querystats.txt)
The queries are the ones ran originally, not simplified for raising issue, but the same issue exists.

You seem to have misunderstood and replicated. "Filter: trans.transdate > custlist.min_trans_date" is not what I saw (which is the expected one). I saw "Filter: trans.bank > custlist.min_trans_date" 

Also, as I mentioned, the issue does not come up when the group by clause is removed.

I am not using the enterprise edition, but the open source edition. ( greenplum-db-4.3.10.0-build-1-rhel5-x86_64.zip https://network.pivotal.io/api/v2/products/pivotal-gpdb/releases/2684/product_files/8431/download)

**Comment by kuruv** [comment]
@vraghavan78 , is this going to be taken up?

**Comment by hlinnaka** [comment]
The incorrect Filter being printed in the EXPLAIN output was fixed in commit c38c72ff979bf4a7e618b5a19700df8cc6d67159. It's now not printed at all, because it was misleading and sometimes outright wrong, like in this query.

Is there something left to do for the out of VMEM issue here?

**Comment by berlin-ab** [comment]
@vraghavan78 Can we triage this issue? Is it still an issue? If not, can we close it?
